### PR TITLE
feat(spec): declare the basic catalog functions and add clamp, round, min, max and abs, add live-calculator example ( fixes #302 )

### LINKED ISSUE #302: Ability to declare client-side event handling where one component can affect another component
Since I, in parallel, worked on a (much smaller) project that also returns agent-generated UI specs based on user prompts, I want to state the following proposal for discussion (and possibly refinement, if not outright-rejected). I purposely keep it relatively vague without going into implementation details in order to keep it short & the discussion on a high level initially, since it may introduce principle target use case incompatibilities with A2UI and thus cause a "desk reject", not warranting deeper investigation.

# Use Case

As a _developer_, I want to be able to enable users to "live" preview the results of their UI interactions without requiring re-prompting the LLM for efficiency and UX, at least for simple transformations/calculations.

As a _user_, I want to be able to make fine-adjustments within the generated UI and receive feedback on them without delays. E.g. in an LLM-based color manipulator.

```
1. Initial prompt: "dark red"
2. Output: Multi-choice selector with color previews of 3 "dark red" variants
<user selects variant A>
3. Second prompt: "make it more saturated"
4. Output: Slider with saturation adjustment and live preview of saturation change of the chosen color (variant A) // <-- this requires either a specialized component, **or a generated callback function that is executed on the client-side**
<user plays around with the saturation slider and receives a live preview of the result>
```

# Current limitations

The current protocol is limited to pre-defined components with built-in business logic (e.g. color saturation change of color in the above example) limiting user-defined actions to strongly pre-defined set OR a whole server<->client roundtrip with A2A messages etc. introducing potentially significant delays, especially as soon as the LLM is re-involved.  

# Extension Concept

1. Enable the (LLM) agent to generate expressions or functions with widget-specific parameters (e.g. slider min, max, position) by introducing the required protocol extensions, e.g. based on a python slang or pure math expressions
2. The functions are executed as a callback to the respective UI component's action (e.g. a slider and a slider position update) client-sided in a sandboxed logic interpreter / expression evaluator, such as python's `asteval` or `simpleeval` or typescript `filtrex` or similar
3. The respective UI component is updated purely client-sided with the function result (e.g. the color preview)

## Benefits

This allows for much more flexible UIs with instant, client-side-computed previews, etc. which are particularly important for fine-tuning results, such as small percentage adjustments in relative changes (for colors: hue, saturation, vibrance, ..) or non-conforming input formats (e.g. a color preview requiring hex input but the user providing rgba in floating point).

## Potential Risks
1. Security issues: expression/script evaluators come with the risk of suffering from security issues, e.g. `expr-eval` with [CVE-2025-12735](https://nvd.nist.gov/vuln/detail/CVE-2025-12735)
2. Incompatibility with the target use case of the A2UI protocol
3. ...?

---

I am looking forward to your feedback, hopefully a discussion and will try to answer all potential questions.

### PR DESCRIPTION
**Pull request by ClemensSchartmueller**
# Description

## Summary

Closes #302, which asks for a way to declare client-side interactions where one
component affects another without a server roundtrip — the motivating case being
a slider that updates a preview immediately rather than re-prompting the model.

This needs no new protocol capability. The reactive machinery already exists:
bindings are re-evaluated when the data they read changes, and catalog functions
are invoked by name against a client-trusted implementation. What was missing was
that most of those functions were never *declared* in the basic catalog, so an
agent had no way to discover them and a strict catalog validator would reject
them. This PR closes that gap, fills in the numeric helpers the pattern needs,
and documents it.

Nothing executable crosses the wire — only catalog-declared function names whose
arguments are validated by JSON Schema — so the expression-evaluator risk raised
in the issue does not apply.

Instead of allowing a generic execution-lib, this declares (as previously existing) explicit functions and re-implements them to avoid (extenral) RCE issues. Additionally, this has the benefit of preserving the existing spec beyond what is mentioned in section **Breaking changes**.

## What changed

**Specification.** `arithmetic`, `comparison` and `string` functions are now
declared in `catalogs/basic/catalog.json`; they were implemented all along but
undeclared. Five new numeric functions are added: `clamp`, `round`, `min`, `max`
and `abs`. `docs/a2ui_protocol.md` gains a section on reactive client-side
computation, and `37_live-calculator.json` demonstrates the full loop with two
`Slider`s driving `Text` values through `add`, `multiply`, `divide`, `clamp` and
`formatCurrency`. All of this lands in `v0_9`, `v0_9_1` and `v1_0`, with the
`v0_9`/`v0_9_1` catalogs and test cases staying byte-identical and `v1_0` keeping
its `FunctionCommon` wrapper and sibling `returnType`. Every specification file
in this PR is a pure addition — no existing schema entry, description or test
case was modified.

**web_core.** Implements the five functions for `v0_9`. `round` rounds halfway
cases away from zero (`-2.5` → `-3`, not `Math.round`'s `-2`) and shifts the
decimal exponent instead of multiplying by a power of ten, so `round(1.005, 2)`
is `1.01` rather than the `1.00` that binary representation error would give.
`clamp` returns `min` when `max` is below `min`, as the catalog declares.

**a2ui_core.** Mirrors those implementations in Python, using `Decimal` with
`ROUND_HALF_UP` to match the away-from-zero tie behaviour that the built-in
`round`'s banker's rounding does not provide. The API classes for the arithmetic,
comparison and string functions are now generated from the catalog schemas rather
than hand-written, which removes the duplication that let the two drift apart.

## Breaking changes

**`divide` by zero now follows IEEE 754.** The implementation special-cased a
zero divisor and returned positive `Infinity` unconditionally, which contradicted
the catalog description already on `main`. The sign of the dividend now carries,
so `-1 / 0` is `-Infinity` and `0 / 0` is `NaN`. `1 / 0` is unchanged. This is
the implementation moving to the spec rather than the spec moving, but it does
change rendered output for anyone who depended on the old value.

**`a2ui.core.basic_catalog.operator_apis` is removed.** Its classes are generated
into `function_apis` and re-exported from `a2ui.core.basic_catalog` under the
same names with identical argument types, so importing from the package root is
unaffected; only a deep import of the submodule path breaks. `equals` and
`not_equals` keep untyped `a`/`b` arguments, matching the `z.any()` contract on
the TypeScript side, so `null` comparisons and exact integer comparisons behave
as before.

Both are recorded in the respective `CHANGELOG.md` files.

## Test coverage

New unit tests cover the five functions in both `web_core` and `a2ui_core`,
including the tie-rounding and `max < min` edge cases, plus the corrected
divide-by-zero results. `function_catalog_validation.json` exercises the newly
declared function schemas across all three spec versions, and the
`37_live-calculator` example has renderer specs in Angular, Lit and React.

### Angular

https://github.com/user-attachments/assets/6193a34f-053f-45ad-80a7-e442732827a4

### Lit

https://github.com/user-attachments/assets/005aa7e9-8232-48c2-bde5-af0f7ea35541

### React

https://github.com/user-attachments/assets/a0a0ebcf-709b-4868-94d6-257fbbaa4bb6

## Pre-launch Checklist

One time:

- [x] I signed the [CLA].
- [x] I read the [Contributors Guide].
- [x] I read the [Style Guide].
- [x] I have built and run at least one [sample].

For this PR:

- [x] I have updated the relevant CHANGELOG.md file.
- [x] I updated/added relevant documentation.
- [x] My code changes (if any) have tests.
- [x] If my branch is on a fork, I have verified that [scripts/e2e_test.sh](https://github.com/a2ui-project/a2ui/blob/main/scripts/e2e_test.sh) passes.

If you need help, consider asking for advice on the [discussion board].

[CLA]: https://cla.developers.google.com/
[Contributors Guide]: https://github.com/a2ui-project/a2ui/blob/main/CONTRIBUTING.md
[discussion board]: https://github.com/a2ui-project/a2ui/discussions
[Style Guide]: https://github.com/a2ui-project/a2ui/blob/main/CONTRIBUTING.md#coding-style
[sample]: https://github.com/a2ui-project/a2ui/blob/main/samples/README.md#samples

### CODE REVIEW

**Comment by ClemensSchartmueller** [comment]
I appreciate intermediate feedback, which is why i created this "WiP" PR.

File: agent_sdks/python/a2ui_core/src/a2ui/core/basic_catalog/function_impls.py [hunk 1]
```diff
@@ -548,6 +552,90 @@ def _ends_with_execute(
...
+def _clamp(
+    args: Dict[str, Any],
+    context: Any = None,
+    abort_signal: Optional[Any] = None,
+) -> Union[int, float]:
+    value = _to_float(args["value"])
+    minimum = _to_float(args["min"])
+    maximum = _to_float(args["max"])
+    # The catalog declares that a maximum below the minimum yields the minimum.
+    res = minimum if maximum < minimum else min(max(value, minimum), maximum)
```
[the remark this answers is not in the document]
Comment (ClemensSchartmueller) [reply agent_sdks/python/a2ui_core/src/a2ui/core/basic_catalog/function_impls.py:564]: this is pre-validated by functions.py via `self.schema.-model.validate...` so no need to validate again?

File: agent_sdks/python/a2ui_core/src/a2ui/core/basic_catalog/function_impls.py [hunk 2]
```diff
@@ -548,6 +552,90 @@ def _ends_with_execute(
...
+    # what web_core does and what the built-in round() does not: round(2.5) is 2
+    # and round(-2.5) is -2 under banker's rounding.
+    #
+    # Quantizing a Decimal built from str(value) also keeps the decimal literal
+    # the user wrote rather than its binary approximation, so round(1.005, 2) is
+    # 1.01 rather than the 1.00 that scaling a float would give.
+    try:
+        quantized = Decimal(str(value)).quantize(
+            Decimal(1).scaleb(-decimals), rounding=ROUND_HALF_UP
+        )
```
[the remark this answers is not in the document]
Comment (ClemensSchartmueller) [reply agent_sdks/python/a2ui_core/src/a2ui/core/basic_catalog/function_impls.py:591]: scaleb should handle large negative factors via InvalidOperatio nexcaptions and clamping would probably introduce be some unchecked / unintended behavior that may alter what consumers of the API see vs. what they expect?

File: renderers/web_core/src/v0_9/basic_catalog/functions/basic_functions.ts [hunk 3]
```diff
@@ -84,12 +91,71 @@ export const DivideImplementation = createFunctionImplementation(DivideApi, args
...
+ *
+ * Multiplying reintroduces the binary representation error that rounding is
+ * meant to hide: `1.005 * 100` is `100.49999999999999`, which rounds down to
+ * `1.00`. `Number('1.005e2')` is `100.5`, which rounds to the expected `1.01`.
+ */
+function shiftExponent(value: number, exponent: number): number {
+  const [mantissa, currentExponent] = String(value).split('e');
+  const shifted = currentExponent ? Number(currentExponent) + exponent : exponent;
+  return Number(`${mantissa}e${shifted}`);
+}
```
[the remark this answers is not in the document]
Comment (ClemensSchartmueller) [reply renderers/web_core/src/v0_9/basic_catalog/functions/basic_functions.ts:119]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toString only specifies lower-case e notation ? so this shouldnt be happening.. ever?

File: renderers/web_core/src/v0_9/basic_catalog/functions/basic_functions.ts [hunk 4]
```diff
@@ -84,12 +91,71 @@ export const DivideImplementation = createFunctionImplementation(DivideApi, args
...
+ * Halfway cases round away from zero, so -2.5 becomes -3 rather than the -2
+ * that JavaScript's Math.round would produce. Non-finite values, and shifts
+ * large enough to overflow to Infinity, are returned unchanged.
+ */
+export const RoundImplementation = createFunctionImplementation(RoundApi, args => {
+  const decimals = args.decimals ?? 0;
+  if (!Number.isFinite(args.value)) return args.value;
+  const scaled = shiftExponent(Math.abs(args.value), decimals);
+  if (!Number.isFinite(scaled)) return args.value;
+  return Math.sign(args.value) * shiftExponent(Math.round(scaled), -decimals);
```
[the remark this answers is not in the document]
Comment (ClemensSchartmueller) [reply renderers/web_core/src/v0_9/basic_catalog/functions/basic_functions.ts:139]: there shouldnt be any performance issues since it the decimals string doesnt really grow meaningfully due to scientific notation ?

**Comment by ClemensSchartmueller** [comment]
for me this seems ready. any objections ? I'll obviously fix the conflict with CHANGELOG.md as soon as there are no further review comments.

**Comment by ClemensSchartmueller** [comment]
do you need any updates on this? i noticed the lint issues (beyond the merge conflicts)

**Comment by polina-c** [comment]
FYI, @jacobsimionato

**Comment by jacobsimionato** [comment]
@ClemensSchartmueller thank you for preparing this PR, though we would prefer not to add this functionality to the basic catalog, so I will close this PR.

What you've built here is compelling and satisfies real use cases.  However, at this stage, we would like to keep the basic catalog as basic as possible, to maintain it as an illustrative catalog that others can extend or be inspired by. Adding more functions to it sets up an expectation that we will expand the set of functions indefinitely with no clear scope. While this isn't too difficult to implement in one or two frameworks, it raises the cost of creating a new basic catalog implementation in a new language etc.

Side note: we cannot change the v0.9 basic catalog because it's already released. v1.0 is in progress, but we are also avoiding new changes in the interests of getting it implemented as soon as possible.

I'd recommend defining a new catalog with this functionality and publishing it as a separate package on NPM etc! We are very curious what new catalogs the A2UI community might invent. A new catalog could depend on the basic catalog and reuse basic catalog functions and components, by pulling them in via the relevant NPM packages etc.
